### PR TITLE
Add task app.tree

### PR DESCRIPTION
**Pull request by liveforeverx**
Task to show dependency tree of an application.

### CODE REVIEW

**Comment by josevalim** [comment]
Thank you!  :heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

File: lib/mix/lib/mix/tasks/app.tree.ex [hunk 1]
```diff
@@ -0,0 +1,70 @@
+defmodule Mix.Tasks.App.Tree do
+  use Mix.Task
+
+  @shortdoc "Analyses the current Mix project and prints dependency tree on terminal."
```
Comment (lexmag) [inline lib/mix/lib/mix/tasks/app.tree.ex]: Probably we need to remove terminal wording, it's just a particular case when `Mix.shell` is set to `IO`.

File: lib/mix/lib/mix/tasks/app.tree.ex [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  @moduledoc """
+  Analyses the current Mix project and prints dependency tree on terminal.
+
+  ## Command line options
+
+    * `--exclude` - exclude dependencies, which you do not want to see in a tree
+
+  ## Example of use
+
+      mix deps.tree --exclude logger,elixir
```
Comment (lexmag) [inline lib/mix/lib/mix/tasks/app.tree.ex]: I think we should implement `--exclude` option without `String.split/2` but by using `:keep`, that's what we do for `mix test` `--exclude` and `--include` options.

File: lib/mix/lib/mix/tasks/app.tree.ex (hunk 1 above)
Comment (liveforeverx) [inline lib/mix/lib/mix/tasks/app.tree.ex]: Have thought, but do not know, how really reword it. What would be right wording for it?

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Comment by alco** [comment]
I just tried this on master and it explodes in a project that hasn't been compiled:

```
$ mix app.tree
** (CaseClauseError) no case clause matching: {:error, {'no such file or directory', 'hashids.app'}}
    (mix) lib/mix/tasks/app.tree.ex:37: Mix.Tasks.App.Tree.load/1
    (mix) lib/mix/tasks/app.tree.ex:32: Mix.Tasks.App.Tree.deps/2
    (mix) lib/mix/tasks/app.tree.ex:28: Mix.Tasks.App.Tree.run/1
    (mix) lib/mix/cli.ex:58: Mix.CLI.run_task/2
    (elixir) lib/code.ex:363: Code.require_file/2
```

File: lib/mix/lib/mix/tasks/app.tree.ex [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  @moduledoc """
+  Analyses the current Mix project and sends dependency tree to mix shell.
+
+  ## Command line options
+
+    * `--exclude` - exclude dependencies, which you do not want to see in a tree
+
+  ## Example of use
+
+      mix deps.tree --exclude logger --exclude elixir
```
Comment (alco) [inline lib/mix/lib/mix/tasks/app.tree.ex]: This should be `app.tree`.

**Comment by josevalim** [comment]
@alco yup, i have fixed it locally :)

File: lib/mix/lib/mix/tasks/app.tree.ex (hunk 3 above)
Comment (liveforeverx) [inline lib/mix/lib/mix/tasks/app.tree.ex]: Yes, sorry overseen by porting...

**Comment by TBK145** [comment]
This is a really cool feature! Will it also be available for umbrella apps?

**Comment by ericmj** [comment]
Shouldn't this be `deps.tree` instead? In which circumstances would `apps.tree` be more useful than `deps.tree`?

**Comment by josevalim** [comment]
We have both in master. app.tree is great to see what we will effectively
depend on in production. I posted a gist online:
https://gist.github.com/josevalim/67cc4fdbd66f96c647fd

App.tree doesn't work for umbrellas but deps.tree does. I will fix that.


-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

**Comment by fishcakez** [comment]
Do we always need to show `elixir` in the `apps.tree`? It creates a lot of noise in the output. Of course a dependency may not require elixir so perhaps we could enable it with a flag? We hide `erts`, `kernel` and `stdlib` - though it is not quite the same brcause I'm not sure it possible to not require those 3 apps, whereas `elixir` is not always required.

**Pull request merged**